File: mobile/src/main/java/net/mm2d/dmsexplorer/util/Downloader.java
/*
 * Copyright (c) 2017 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.dmsexplorer.util;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import net.mm2d.upnp.Http;
import net.mm2d.upnp.HttpClient;
import net.mm2d.upnp.HttpResponse;

import java.io.IOException;
import java.net.URL;

import io.reactivex.Single;
import io.reactivex.SingleOnSubscribe;
import io.reactivex.schedulers.Schedulers;

/**
 * @author <a href="mailto:ryo@mm2d.net">大前良介 (OHMAE Ryosuke)</a>
 */
public class Downloader {
    private Downloader() {
        throw new AssertionError();
    }

    @NonNull
    public static Single<byte[]> create(@NonNull final String url) {
        return Single.create((SingleOnSubscribe<byte[]>) emitter -> {
            final byte[] binary = download(url);
            if (binary != null) {
                emitter.onSuccess(binary);
            } else {
                emitter.onError(new IOException());
            }
        }).subscribeOn(Schedulers.io());
    }

    @Nullable
    private static byte[] download(@NonNull final String url) {
        try {
            final HttpResponse response = new HttpClient(false).download(new URL(url));
            if (response.getStatus() == Http.Status.HTTP_OK) {
                return response.getBodyBinary();
            }
        } catch (final IOException ignored) {
        } catch (final OutOfMemoryError ignored) {
        }
        return null;
    }
}

File: mobile/src/main/java/net/mm2d/dmsexplorer/util/Downloader.kt
/*
 * Copyright (c) 2017 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.dmsexplorer.util

import io.reactivex.Single
import net.mm2d.upnp.Http
import net.mm2d.upnp.HttpClient
import java.io.IOException
import java.net.URL

/**
 * @author [大前良介 (OHMAE Ryosuke)](mailto:ryo@mm2d.net)
 */
object Downloader {
    @JvmStatic
    fun create(url: String): Single<ByteArray> {
        return Single.create { emitter ->
            val binary = download(url)
            if (binary != null && binary.isNotEmpty()) {
                emitter.onSuccess(binary)
            } else {
                emitter.onError(IOException())
            }
        }
    }

    private fun download(url: String): ByteArray? {
        try {
            val response = HttpClient(false).download(URL(url))
            if (response.status == Http.Status.HTTP_OK) {
                return response.bodyBinary
            }
        } catch (ignored: IOException) {
        } catch (ignored: OutOfMemoryError) {
        }
        return null
    }
}
